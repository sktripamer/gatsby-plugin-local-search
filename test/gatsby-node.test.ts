import { CreatePagesArgs, CreateSchemaCustomizationArgs } from 'gatsby'
import lunr from 'lunr'
import FlexSearch from 'flexsearch'

import { createPages, createSchemaCustomization } from '../src/gatsby-node'
import { PluginOptions } from '../src/types'

const mockQueryResult = {
  data: {
    allNode: {
      nodes: [
        { id: 'id1', foo: 'bar' },
        { id: 'id2', foo: 'needle' },
        // Without valid ref field.
        { id: undefined, foo: 'baz' },
      ],
    },
  },
} as const

const pluginOptions: PluginOptions = {
  name: 'name',
  engine: 'flexsearch',
  query: 'query',
  normalizer: (input) =>
    (input as typeof mockQueryResult).data.allNode.nodes.map((node) => ({
      id: node.id,
      foo: node.foo,
    })),
  plugins: [] as unknown[],
}

beforeEach(() => jest.clearAllMocks())

describe('createPages', () => {
  const mockGatsbyContext: CreatePagesArgs & {
    traceId: 'initial-createPages'
  } = {
    // @ts-expect-error Partial actions
    actions: { createNode: jest.fn() },
    // @ts-expect-error Partial reporter
    reporter: { warn: jest.fn(), error: jest.fn() },
    createNodeId: jest.fn().mockReturnValue('createNodeId'),
    createContentDigest: jest.fn().mockReturnValue('createContentDigest'),
    graphql: jest.fn().mockReturnValue(Promise.resolve(mockQueryResult)),
  }

  describe('with flexsearch engine', () => {
    test('creates index', async () => {
      await createPages(mockGatsbyContext, pluginOptions)

      const createNode = mockGatsbyContext.actions.createNode as jest.Mock
      const node = createNode.mock.calls[0][0]
      const exportedIndex = node.index
      const index = FlexSearch.create()
      index.import(exportedIndex)

      expect(mockGatsbyContext.actions.createNode).toMatchSnapshot()
      expect(index.search('needle')).toMatchSnapshot()
    })

    test('passes engine options', async () => {
      const spy = jest.spyOn(FlexSearch, 'create')
      const engineOptions = { rtl: true }

      await createPages(mockGatsbyContext, { ...pluginOptions, engineOptions })

      expect(spy).toHaveBeenCalledWith(engineOptions)
      spy.mockRestore()
    })
  })

  describe('with lunr engine', () => {
    test('creates index', async () => {
      await createPages(mockGatsbyContext, { ...pluginOptions, engine: 'lunr' })

      const createNode = mockGatsbyContext.actions.createNode as jest.Mock
      const node = createNode.mock.calls[0][0]
      const exportedIndex = node.index
      const index = lunr.Index.load(JSON.parse(exportedIndex))

      expect(mockGatsbyContext.actions.createNode).toMatchSnapshot()
      expect(index.search('needle')).toMatchSnapshot()
    })
  })
})

describe('createSchemaCustomization', () => {
  const mockGatsbyContext: CreateSchemaCustomizationArgs = {
    // @ts-expect-error Partial actions
    actions: { createTypes: jest.fn() },
    // @ts-expect-error Partial schema
    schema: {
      buildObjectType: jest
        .fn()
        .mockImplementation((config) => ({ kind: 'OBJECT', config })),
    },
  }

  test('creates types', async () => {
    await createSchemaCustomization(mockGatsbyContext, pluginOptions)

    expect(mockGatsbyContext.actions.createTypes).toMatchSnapshot()
  })
})
