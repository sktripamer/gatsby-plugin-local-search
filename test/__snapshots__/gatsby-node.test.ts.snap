// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPages with flexsearch engine creates index 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "engine": "flexsearch",
        "id": "createNodeId",
        "index": "[[{\\"i\\":[\\"id1\\",\\"id2\\"],\\"id\\":[\\"id1\\",\\"id2\\"]},{\\"id1\\":[\\"id1\\"],\\"f\\":[\\"id1\\",\\"id2\\"],\\"fo\\":[\\"id1\\",\\"id2\\"],\\"foo\\":[\\"id1\\",\\"id2\\"],\\"id2\\":[\\"id2\\"]},{\\"b\\":[\\"id1\\"],\\"ba\\":[\\"id1\\"],\\"bar\\":[\\"id1\\"],\\"n\\":[\\"id2\\"],\\"ne\\":[\\"id2\\"],\\"nee\\":[\\"id2\\"],\\"need\\":[\\"id2\\"],\\"needl\\":[\\"id2\\"],\\"needle\\":[\\"id2\\"]},{},{},{},{},{},{}],{},[\\"@id1\\",\\"@id2\\"]]",
        "internal": Object {
          "contentDigest": "createContentDigest",
          "type": "LocalSearchName",
        },
        "name": "name",
        "store": Object {
          "id1": Object {
            "foo": "bar",
            "id": "id1",
          },
          "id2": Object {
            "foo": "needle",
            "id": "id2",
          },
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`createPages with flexsearch engine creates index 2`] = `
Array [
  "id2",
]
`;

exports[`createPages with lunr engine creates index 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "engine": "lunr",
        "id": "createNodeId",
        "index": "{\\"version\\":\\"2.3.8\\",\\"fields\\":[\\"id\\",\\"foo\\"],\\"fieldVectors\\":[[\\"id/id1\\",[0,0.693]],[\\"foo/id1\\",[1,0.693]],[\\"id/id2\\",[2,0.693]],[\\"foo/id2\\",[3,0.693]]],\\"invertedIndex\\":[[\\"bar\\",{\\"_index\\":1,\\"id\\":{},\\"foo\\":{\\"id1\\":{}}}],[\\"id1\\",{\\"_index\\":0,\\"id\\":{\\"id1\\":{}},\\"foo\\":{}}],[\\"id2\\",{\\"_index\\":2,\\"id\\":{\\"id2\\":{}},\\"foo\\":{}}],[\\"needl\\",{\\"_index\\":3,\\"id\\":{},\\"foo\\":{\\"id2\\":{}}}]],\\"pipeline\\":[\\"stemmer\\"]}",
        "internal": Object {
          "contentDigest": "createContentDigest",
          "type": "LocalSearchName",
        },
        "name": "name",
        "store": Object {
          "id1": Object {
            "foo": "bar",
            "id": "id1",
          },
          "id2": Object {
            "foo": "needle",
            "id": "id2",
          },
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`createPages with lunr engine creates index 2`] = `
Array [
  Object {
    "matchData": Object {
      "metadata": Object {
        "needl": Object {
          "foo": Object {},
        },
      },
    },
    "ref": "id2",
    "score": 0.693,
  },
]
`;

exports[`createSchemaCustomization creates types 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "config": Object {
            "fields": Object {
              "engine": Object {
                "description": "The search engine used to create the index.",
                "type": "String!",
              },
              "index": Object {
                "description": "The search index created using the selected engine.",
                "type": "String!",
              },
              "name": Object {
                "description": "The name of the index.",
                "type": "String!",
              },
              "publicIndexURL": Object {
                "description": "Save the index to the site's static directory and return a public URL to it.",
                "resolve": [Function],
                "type": "String!",
              },
              "publicStoreURL": Object {
                "description": "Save the store to the site's static directory and return a public URL to it.",
                "resolve": [Function],
                "type": "String!",
              },
              "store": Object {
                "description": "A JSON object used to map search results to their data.",
                "type": "JSON!",
              },
            },
            "interfaces": Array [
              "Node",
            ],
            "name": "LocalSearchName",
          },
          "kind": "OBJECT",
        },
      ],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
